src/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture.spec.ts

```diff
@@ -150,19 +164,195 @@ class CopyImageBitmapToTextureTest extends ValidationTest {
+      { texture: dstTexture },
+      copySize,
+      true, // No validation errors.
+      isValidContextType(contextType) ? '' : 'OperationError'
+    );
+  });
+
+g.test('source_offscreenCanvas,contexts')
+  .desc(
+    `
+  Test OffscreenCanvas as source image with different contexts.
+
+  Call OffscreenCanvas.getContext() with different context type.
+  Only '2d', 'webgl', 'webgl2' is valid context type.
+
+  Check whether 'OperationError' is generated when context type is invalid.
+  `
+  )
+  .params(u =>
+    u //
+      .combine('contextType', ['2d', 'bitmaprenderer', 'webgl', 'webgl2'] as const)
+      .beginSubcases()
+      .combine('copySize', [
+        { width: 0, height: 0, depthOrArrayLayers: 0 },
+        { width: 1, height: 1, depthOrArrayLayers: 1 },
+      ])
+  )
+  .fn(async t => {
+    const { contextType, copySize } = t.params;
+    const canvas = t.getOffscreenCanvas(1, 1);
+
+    const dstTexture = t.device.createTexture({
+      size: { width: 1, height: 1, depthOrArrayLayers: 1 },
+      format: 'bgra8unorm',
+      usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    canvas.getContext(contextType);
+
+    t.runTest(
+      { source: canvas },
+      { texture: dstTexture },
+      copySize,
+      true, // No validation errors.
+      isValidContextType(contextType) ? '' : 'OperationError'
+    );
+  });
+
+g.test('source_image,crossOrigin')
+  .desc(
+    `
+  Test contents of source image is [clean, cross-origin].
+
+  Load crossOrigin image or same origin image and init the source
+  images.
+  
+  Check whether 'SecurityError' is generated when source image is not origin clean.
+  `
+  )
+  .params(u =>
+    u //
+      .combine('sourceImage', ['canvas', 'offscreenCanvas', 'imageBitmap'])
+      .combine('isOriginClean', [true, false])
+      .beginSubcases()
+      .combine('copySize', [
+        { width: 0, height: 0, depthOrArrayLayers: 0 },
+        { width: 1, height: 1, depthOrArrayLayers: 1 },
+      ])
+  )
+  .fn(async t => {
+    const { sourceImage, isOriginClean, copySize } = t.params;
+
+    const crossOriginUrl = 'https://get.webgl.org/conformance-resources/opengl_logo.jpg';
+    const originCleanUrl = '../out/resources/Di-3d.png';
+
+    const img = document.createElement('img');
+    img.src = isOriginClean ? originCleanUrl : crossOriginUrl;
+    try {
+      await awaitOrTimeout(img.decode());
+    } catch (e) {
+      t.skip('Cannot load image');
+    }
+
+    const canvas =
+      sourceImage === 'offscreenCanvas'
+        ? t.getOffscreenCanvas(1, 1)
+        : document.createElement('canvas');
+
+    const ctx = canvas.getContext('2d');
+    if (ctx === null) {
+      t.skip('Cannot get 2d context');
+    } else {
+      ctx.drawImage(img, 0, 0);
+    }
+
+    let externalImage: HTMLCanvasElement | OffscreenCanvas | ImageBitmap = canvas;
+    if (sourceImage === 'imageBitmap') {
+      externalImage = await createImageBitmap(canvas);
```

Comment:
I was thinking the imagebitmap case would be creating the imagebitmap directly from the image, rather than going through canvas. But actually, we should test both!